### PR TITLE
Change default top_k to 50 everywhere for consistency

### PR DESCRIPTION
**Pull request by rasbt**
If the current default top_k values we sometimes get really bad answers, which may make LitGPT look really bad. Just an example:

```
from litgpt import LLM

llm = LLM.load("mistralai/Mathstral-7B-v0.1")
print(llm.generate("What is 1+2?"))
```

```
 In your job, what is the average time people spend using your app? These can have wildly different answers depending upon the dataset, the timeframe for the dataset, and how the data is being calculated or stated, disaggregated and re
```

With 

```
print(llm.generate("What is 1+2?", top_k=1))
```

```
 1+2 equals 3.
```

Overall, I think it's better to err on the side of generating the same answer, and if people want more variation, they can change the `top_k` setting. I think this is better than generating garbage by default and people trying to figure out how to make it work properly.

### CODE REVIEW

File: litgpt/api.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ def generate(
             prompt: Tensor of shape (T) with indices of the prompt sequence.
             max_returned_tokens: The maximum number of tokens to return (given plus generated).
             temperature: Scales the predicted logits by 1 / temperature.
-            top_k: If specified, only sample among the tokens with the k highest probabilities.
+            top_k: Only sample among the tokens with the k highest probabilities.
```
Comment (rasbt) [inline litgpt/api.py:200]:
```suggestion
            top_k: If specified, only sample among the tokens with the k highest probabilities.
```

File: litgpt/api.py [hunk 2]
```diff
@@ -183,7 +183,7 @@ def generate(
         prompt: str,
         max_new_tokens: int = 50,
         temperature: float = 1.0,
-        top_k: Optional[int] = None,
+        top_k: Optional[int] = 1,
```
Comment (rasbt) [inline litgpt/api.py:186]:
```suggestion
        top_k: Optional[int] = None,
```

**Comment by Andrei-Aksionov** [comment]
While I agree with the reasoning, I'm concerned that this happens in the first place.

Let's merge the PR and later come back to it.
Current default values are quite safe and that shouldn't happen, in theory.

**Comment by awaelchli** [comment]
@rasbt How do you arrive at the conclusion that top_k needs to be set to 1 in this case? What do the other outputs look like under top_k > 1 and does the original model also exhibit this behavior? What about the other models (e.g., see other [references](https://github.com/meta-llama/llama/blob/be327c427cc5e89cc1d3ab3d3fec4484df771245/example_text_completion.py#L13C12-L13C24))?

**Comment by rasbt** [comment]
It's basically something observed at the workshop last week too, it takes several takes sometimes to get reasonable results, which looks kind of bad. Here for phi-2:

Screenshot 2024-07-18 at 11 39 25 AM

For `top_k=1` you will always get

> ' Llamas are herbivores and primarily eat grass, hay, and other plant material. They have a unique digestive system that allows them to efficiently extract nutrients from fibrous plant material.\n'

Other values like `top_k=5` or 10 also works ok. But isn't it better to make it deterministic by default to ensure somewhat  good results? Either way people may want to change the setting, but I think it's better if it looks good by default then producing weird results and people thinking it's an error, issue or bug.

**Comment by rasbt** [comment]
Looks like the mileage varies depending on what model we look at, but I think that 50 is definitely too high. Would something like 5 be a good compromise where we have it not strictly deterministic but also not quite as random? What do you think @awaelchli @Andrei-Aksionov ?

**Comment by Andrei-Aksionov** [comment]
Could you then also create an issue with these findings?

Definitely something is off.

**Review by Andrei-Aksionov — APPROVED** [review]

**Comment by Andrei-Aksionov** [comment]
Yes, let's stick to 50.

Create an issue.
I'll take a look at it when I have time.

**Comment by rasbt** [comment]
Ok sounds good. In the meantime, in this PR I changed everything to 50 then. It was 200 in some places.

**Comment by awaelchli** [comment]
What I don't understand is how the conclusion from "model does not produce good outputs" can be to reduce top_k. If top_k is really so impactful, shouldn't this be a problem in other implementations as well? 

For example:
```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

torch.set_default_device("cuda")

model = AutoModelForCausalLM.from_pretrained("microsoft/phi-2", torch_dtype="auto", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained("microsoft/phi-2", trust_remote_code=True)

inputs = tokenizer("What do Llamas eat?", return_tensors="pt", return_attention_mask=False)

outputs = model.generate(**inputs, max_length=200, top_k=200, do_sample=True)
text = tokenizer.batch_decode(outputs)[0]
print(text)
```
produces various examples of similar quality as what you have shared as acceptable. 
Maybe I'm doing something wrong here on the HF transformers side to make this comparison. Please correct me if I'm wrong.

**Comment by rasbt** [comment]
@awaelchli  I think I got bad results before but never that bad. I think it's possibly also something related to the kv-cache but not sure. I opened an issue here #1593

**Comment by awaelchli** [comment]
This hints at an issue, either in my transformers comparison code, or in LitGPT's generation code. If the latter is true, setting topk=1 should be considered a temporary hack in my opinion.

**Review by awaelchli — APPROVED** [review]

**Comment by rasbt** [comment]
What makes me suspicious now is the the generation between `litgpt chat` and the Python API look different. The one from the Python API looks much worse. It can't be a total bug because otherwise lower `top_k` values wouldn't work. I need to investigate more. Thanks for calling this out though, it didn't occur to me that it could be anything other then `top_k` setting but now I realize that it might be something different.

**Pull request merged**